doc/design/design_high-available_sqlflow_server.md

```diff
@@ -0,0 +1,231 @@
+
+The `JobRunner` interface should provide two functions `run` and `fetch`:
+
+```go
+type JobRunner interface {
+  run(sql string, pr *PipeReader, pw *PipeWriter) (token string, err error){
+  fetchResult(token string) (result *pb.Result)
+}
+```
+
+Registe `JobRunner` in `sql.Server`:
+
+```go
+func (s *Server) Run(ctx context.Context, req *pb.Request) (*pb.Token, error) {
+  db := s.db
+  pr, pw := sf.Pipe()
+  token := s.jobRunner.run(req.Sql, pr, pw)
+  return &pb.Token(Token: token), nil
+}
+
+func (s *Server) Fetch(ctx context.Context, token *pb.Token) (*pb.Result, error) {
+  result, error := s.jobRunner.fetch(token.Token)
+  return result, nil
+}
+
+func main() {
+  // registe `LocalJobRunner` or `ClusterJobRunner` according to the env variable `SQLFLOW_JOB_RUNNER`
+  server.RegisteJobRunner(os.getenv("SQLFLOW_JOB_RUNNER"))
+}
+```
+
+### Local Job Runner
+
+Upon processing a `Run` request, the server generates, bookkeeps, and returns the token to the client. Upon processing a `Fetch` request, the server looks up the result channel and returns the most recent result.
+
+```go
+type LocalJobRunner {
+  jobs map[string]*PipeReader
+}
+
+func (r *LocalJobRunner) run (sql string, pr *PipeReader, pw *PipeWriter) (string, error){
+  token := tokenGen()
+  r.jobs[token] = pr
+  go func() {
+      defer pw.Close()
+      pw.Write(`RUNNING`)
+      sqlStatements, _ := sf.SplitMultipleSQL(req.Sql)
+      for _, singleSQL := range sqlStatements {
+         for e := range s.run(singleSQL, db, s.modelDir, req.Session).ReadAll() {
+            pw.Write(e)
+         }
+      }
+      pw.Write(`SUCCEEDED`)
+  }()
+  return token
+}
+
+func (r *LocalJobRunner) fetch(token string) (*pb.Result, error) (
+  result := &pb.Result{}
+  pr, ok := r.jobs[token]
+  if !ok {
+      return nil, fmt.Errorf("unrecognized token %s", token.Token)
+   }
+   for ;; {
+      select {
+      case res := <-pr.ReadAll():
+         // construct result
+      case <-time.After(1 * time.Second):
+         break;
+      }
+   }
+   return result, nil
+)
+
+```
+
+Since we have multiple gRPC calls for a server instance, we need to maintain a state across different calls.
+So we use a map `map[string]*PipeReader` to maintain the job states on the server
+
+### Cluster Job Runner on Server
+
+Upon processing a `Run` request, the server launches a Kubernetes Pod and return the Pod ID and log view URL to the client.
+Upon processing a `Fetch` request, the server checks the Pod status and returns the `JobStatus`.
+
+```go
+type ClusterJobRunner {
+}
+
+func (r *LocalJobRunner)run(sql string, pr *PipeReader, pw *PipeWriter) (string, error){
+  podID, err := r.launchK8sPod(sql)
+  pw.Write(fmt.Sprintf("Logs viewer URL: %s", r.logsViewerURL(podID)))
+  return podID, nil
+}
+
+func (r *LocalJobRunner) fetch(token string) (*pb.Result, error) (
+  result := &pb.Result{}
+  result.job_status := r.PodStatus(token)
+  return result, nil
+)
+```
+
+### Store the Trained Model
```

Comment:
Is there a tight relationship between job ID and storing trained models? I am not sure if we should have this section in this document about job ID.

Reply:
Hi @wangkuiyi, you are right. This section is a leftover between @Yancey1989 and me. We were discussing how the Kubernetes jobs should save the trained model.

@Yancey1989 Shall we save this section for future discussions?